File: week03/practice/http-2/main.js
const http = require("http");
const { URL } = require("url");

const PORT = Number(process.env.PORT) || 8080;
const TODOS_URL = "https://jsonplaceholder.typicode.com/todos";

let todosCache = null;

async function getTodos() {
  if (todosCache) {
    return todosCache;
  }

  const response = await fetch(TODOS_URL);
  if (!response.ok) {
    throw new Error(`request failed: ${response.status} ${response.statusText}`);
  }

  const todos = await response.json();
  todosCache = todos.map((todo) => ({
    title: todo.title,
    userId: todo.userId,
    id: todo.id,
    completed: todo.completed,
  }));
  return todosCache;
}

function sendJson(res, statusCode, payload) {
  res.writeHead(statusCode, { "Content-Type": "application/json; charset=utf-8" });
  res.end(JSON.stringify(payload));
}

const server = http.createServer(async (req, res) => {
  try {
    const url = new URL(req.url, `http://${req.headers.host}`);
    const pathname = url.pathname;

    if (req.method === "GET" && pathname === "/api/todo/list") {
      const todos = await getTodos();
      return sendJson(res, 200, todos);
    }

    if (req.method === "GET" && pathname.startsWith("/api/todo/detail/")) {
      const todoId = Number(pathname.replace("/api/todo/detail/", ""));
      if (!Number.isInteger(todoId) || todoId <= 0) {
        return sendJson(res, 400, { message: "todoid must be a positive integer" });
      }

      const todos = await getTodos();
      const todo = todos.find((item) => item.id === todoId);
      if (!todo) {
        return sendJson(res, 404, { message: "todo not found" });
      }
      return sendJson(res, 200, todo);
    }

    return sendJson(res, 404, { message: "route not found" });
  } catch (error) {
    return sendJson(res, 500, { message: "internal server error", detail: error.message });
  }
});

server.listen(PORT, () => {
  console.log(`todo server running at http://localhost:${PORT}`);
});

server.on("error", (error) => {
  if (error.code === "EADDRINUSE") {
    console.error(
      `port ${PORT} is already in use. Stop the existing process or run with another port, e.g. PORT=8081 node week03/practice/http-2/main.js`
    );
    process.exit(1);
  }

  console.error("server start failed:", error.message);
  process.exit(1);
});

File: week03/practice/http-2/main.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"strconv"
	"strings"
	"sync"
)

const (
	port     = ":8080"
	todosURL = "https://jsonplaceholder.typicode.com/todos"
)

type Todo struct {
	Title     string `json:"title"`
	UserID    int    `json:"userId"`
	ID        int    `json:"id"`
	Completed bool   `json:"completed"`
}

var (
	todosCache []Todo
	cacheMu    sync.RWMutex
)

func fetchTodos() ([]Todo, error) {
	cacheMu.RLock()
	if len(todosCache) > 0 {
		copied := make([]Todo, len(todosCache))
		copy(copied, todosCache)
		cacheMu.RUnlock()
		return copied, nil
	}
	cacheMu.RUnlock()

	resp, err := http.Get(todosURL)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(resp.Body)
		return nil, fmt.Errorf("request failed: %d %s", resp.StatusCode, strings.TrimSpace(string(body)))
	}

	var todos []Todo
	if err = json.NewDecoder(resp.Body).Decode(&todos); err != nil {
		return nil, err
	}

	cacheMu.Lock()
	todosCache = todos
	cacheMu.Unlock()

	return todos, nil
}

func writeJSON(w http.ResponseWriter, status int, payload any) {
	w.Header().Set("Content-Type", "application/json; charset=utf-8")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(payload)
}

func todoHandler(w http.ResponseWriter, r *http.Request) {
	if r.Method != http.MethodGet {
		writeJSON(w, http.StatusMethodNotAllowed, map[string]string{"message": "method not allowed"})
		return
	}

	path := r.URL.Path

	if path == "/api/todo/list" {
		todos, err := fetchTodos()
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"message": err.Error()})
			return
		}
		writeJSON(w, http.StatusOK, todos)
		return
	}

	if strings.HasPrefix(path, "/api/todo/detail/") {
		idStr := strings.TrimPrefix(path, "/api/todo/detail/")
		todoID, err := strconv.Atoi(idStr)
		if err != nil || todoID <= 0 {
			writeJSON(w, http.StatusBadRequest, map[string]string{"message": "todoid must be a positive integer"})
			return
		}

		todos, err := fetchTodos()
		if err != nil {
			writeJSON(w, http.StatusInternalServerError, map[string]string{"message": err.Error()})
			return
		}

		for _, todo := range todos {
			if todo.ID == todoID {
				writeJSON(w, http.StatusOK, todo)
				return
			}
		}

		writeJSON(w, http.StatusNotFound, map[string]string{"message": "todo not found"})
		return
	}

	writeJSON(w, http.StatusNotFound, map[string]string{"message": "route not found"})
}

func main() {
	http.HandleFunc("/api/todo/list", todoHandler)
	http.HandleFunc("/api/todo/detail/", todoHandler)

	fmt.Println("todo server running at http://localhost:8080")
	if err := http.ListenAndServe(port, nil); err != nil {
		fmt.Printf("server start failed: %v\n", err)
	}
}
